[BUG] cppclass overloaded alternatives are dropped

BlindChickens opened the pull request:
Fixes https://github.com/cython/cython/issues/4524

BlindChickens commented:
@scoder  After all the tests failed with the first attempt at fixing the bug, I pushed another approach which also resolves the problem for me at least locally.
Can someone please re-approve the workflows to run.

scoder commented on Cython/Compiler/Symtab.py:
```diff
@@ -2622,6 +2622,7 @@ def declare_inherited_cpp_attributes(self, base_class):
                 base_entry.type, None, 'extern')
             entry.is_variable = 1
             entry.is_inherited = 1
+            entry.overloaded_alternatives = base_entry.overloaded_alternatives
```
This should probably make a copy, in case the entry adds overloads – a case that seems worth testing, e.g. with two subclasses that add different overloads. That also suggests errors tests which you can add to `tests/errors/` with a `# mode: error`. See the other tests there on how to list the error output lines.

BlindChickens replied:
Hi @scoder . I have updated my testcase in `tests/run/cpp_class_overloaded_traits.pyx`. As it stands, that test case works with the fix I made.

I don't completely understand what you mean by your suggestion. Can you give me just a small example based on my testcase that you want me to test if it works.

And the error tests you mentioned, also based on the sample test I have here on the PR, if you can just provide a snippet of something you reckon should fail, I will test it, and add a testcase.

Thanks in advance.

da-woods replied:
Something like:

```
cdef:
    struct S1:
        pass

    struct S2:
        pass

	extern from "cpp_source.h":
		cppclass testclass:
			testclass()
            void a()
			void a(unsigned int i)

	cppclass testclass2(testclass):
		testclass2()

	cppclass testclass3(testclass):
		testclass3()
		void a(S1 v)

    cppclass testclass4(testclass):
        testclass4()
        void a(S2 v)

cdef S1 s1
cdef S2 s2
testclass4().a(s1)  # should fail
testclass4().a(s2)  # should pass
testclass3().a(s1)  # should pass
testclass3().a(s2)  # should fail
testclass2().a(s1)  # should fail
testclass2().a(s2)  # should fail
testclass1().a(s1)  # should fail
testclass1().a(s2)  # should fail
```

I think the worry is that because the entries share the same object for `overloaded_alternatives` that if you add an overload to one class then you add them to all classes.

BlindChickens replied:
@da-woods Thanks a lot. I do not really know anything about c, c++ or Cython, so thanks a lot. I think I got everything you and @scoder wanted. Can you have a look and tell me if its alright?

Can we also run the workflows again?

scoder commented on tests/errors/cpp_class_extern_overloaded_traits.pyx:
```diff
@@ -0,0 +1,53 @@
+	struct S2:
+		int i
+
+	extern from "cpp_source.h":
```
You do not need to add this file to `tests/errors/` since the C compiler will never be run after the translation fails in Cython.

scoder commented on tests/run/cpp_class_extern_overloaded_traits.pyx:
```diff
@@ -0,0 +1,49 @@
+	struct S2:
+		int i
+
+	extern from "cpp_source.h":
```
This also doesn't need a separate test file. You can use verbatimly inlined C code.
```suggestion
	extern from *:
	    """
	    #include <stdio.h>
        #include <map>

        unsigned int i;

        void test(std::map<unsigned int, unsigned int> m) {
	        printf("Map size: %d\n", m.size());
        }

        class testclass {
        public:
	        void a() {
		        printf("testclass->a();\n");
	        }
	        void a(unsigned int i) {
		        printf("testclass->a(%d);\n", i);
	        }
        };
	    """
```

scoder commented on tests/run/cpp_class_extern_overloaded_traits.pyx:
```diff
@@ -0,0 +1,49 @@
+# tag: cpp
```
```suggestion
# mode: run
# tag: cpp
```

BlindChickens commented:
@scorder Thanks for the reviews/fixes. I was away for 2 days. I saw the tests fail because of tab characters, I fixed it.

scoder commented:
Can't say if the tests will succeed, but in any case, I suggest to
- change the overloaded C++ functions to return some value instead of calling the low-level `printf()`
- add a Python test function, with a doctest, rather than letting test code execute in the module body
- let the test function return the C++ return value and use it in the doctest to validate that the right function was called.
Otherwise, it will be difficult to debug this test if something goes wrong in the future, because a failure won't provide much of a helpful error output.

BlindChickens commented:
Hi @scoder, I implemented the changes you suggested. I also implemented some changes from this PR https://github.com/cython/cython/pull/3235, because of some of my tests that I eventually saw failed and this looked like it will solve them.

I am however not getting the final piece of the implementation to work. The part where a class would add one overloaded alternative to the inherited method. This code:
```
cppclass testclass3(testclass):
    testclass3()
    void a(S1 v)

cppclass testclass4(testclass):
    testclass4()
    void a(S2 v)
```

I got the code to a point where both these classes would add this method in the `overloaded_alternatives` array, but it just does not want to work when I run the tests. For some reason, it only 'sees' the parameterless method and the first overloaded alternative that takes the `int` as parameter.

I have tried to add this new alternative to the `var_entries` array as well as the alternatives array, I have tried making the method that testclasses 3 and 4 add the primary method in the entries object, and then taking all the previous ones and adding them to the `overloaded_alternatives` of this one, but to no avail.

If I can ask any one of you (@scoder, @da-woods) to just help me out here in a direction. Everything I have done is commited on the PR.

da-woods commented on Cython/Compiler/Symtab.py:
```diff
@@ -2589,12 +2603,9 @@ def maybe_ref(arg):
         entry = self.declare_var(name, type, pos,
                                  defining=defining,
                                  cname=cname, visibility=visibility)
-        if prev_entry and not defining:
-            entry.overloaded_alternatives = prev_entry.all_alternatives()
```
This definitely isn't replaced. So for the top-level `testclass` you don't set up `overloaded_alternatives` for `a`. Instead you just replace `a()` with `a(int)`

BlindChickens replied:
Hi, sorry for the late reply. This is done in `Scope`'s `declare` method. The top-level class works as expected. I can call `a` with nothing or an `int` on `testclass`. I can even add more overload and it works.. But I cannot inherit and then add overload. 

@da-woods Is there any way I can split this PR into my original problem, and then the secondary problem of the inherited class's overload?

So my problem is only the fact that the overloads of an external class got lost when inherited. So basically the fact that `testclass2` would not be able to call `a()` and `a(int)`. But `testclass2` would not add an overload.

da-woods replied:
Hi - sorry if this is a bit vague - I'm struggling to find time for a detailed look at the moment.

On splitting the PR:
* I think it's probably fine to split it into two parts if the smaller PR solves one problem and doesn't actively break anything else.
* My suspicion is that there should be one approach that solves both. However I could be wrong. it's easy for me to say that when I haven't actually tried it.

samaingw commented:
Hello, I just saw this PR, great to see there is some traction around bugfixing/feature addition around the C++ support for Cython.

I read this quite fast, but I suspect this is a subset of #3235, as it seems it targets the case of introducing all the declared alternatives of a base class into its derived classes.
On the other side, #3235 aims to be able to define cppclass methods with alternatives and inheritance working in Cython.

As this PR emerged, I deduce that #3235 is either crap or too complicated to review or with parts that are of no interest.

Therefore, as I hope this PR will get merged (nice contribution @BlindChickens ;) ), I wonder if #3235:
- is interesting but should be split up to help review
- has only some parts which are interesting
- is crap, looking for a better implementation by someone else

@BlindChickens : would you mind if I read a bit your PR and post questions / remarks if I've got any ?
I'd love to help if I can.

khallig commented:
Apologies to be blindly commenting on someone's PR here.

Both this PR: #4525, and #3235 adress an issue with overloading cppclass methods, and fix the issue that I've been consistely experiencing.  It seems that this is the somewhat favoured PR to resolve the issue - I'm basing that off @samaingw's comment above.

It would be fantastic to have this merged down for the next Cython release.

Would the PR's author @BlindChickens or the reviewers know if this is a candidate for being merged down into the next Cython release?
